File: backend/app/core/password_policy.py
"""Shared password strength rule, used by every schema that accepts a new
plaintext password (self-registration, password reset, and admin-side user
create/update). Kept in one place so the rule can never drift between entry
points — the length bound (`Field(min_length=8, max_length=72)`) already
lives on each schema field individually since Pydantic needs it there for
its own error message, but the *complexity* check (uppercase + lowercase +
digit + symbol) is identical everywhere and belongs in one function.

Requires all four character classes — uppercase, lowercase, digit, and a
symbol — on top of the 8-72 char length bound already enforced by each
schema field.
"""

import re

_UPPERCASE_RE = re.compile(r"[A-ZÁÉÍÓÚÑÜ]")
_LOWERCASE_RE = re.compile(r"[a-záéíóúñü]")
_DIGIT_RE = re.compile(r"\d")
# Anything that isn't a letter, digit, or whitespace counts as a symbol —
# deliberately broad so punctuation from any keyboard layout qualifies,
# rather than hardcoding a specific ASCII symbol set.
_SYMBOL_RE = re.compile(r"[^\w\s]", re.UNICODE)


def validate_password_strength(value: str) -> str:
    """Raise ValueError if `value` doesn't contain at least one uppercase
    letter, one lowercase letter, one digit, and one symbol. Returns the
    value unchanged otherwise, so it can be used directly as a Pydantic
    `field_validator` body."""
    if not _UPPERCASE_RE.search(value):
        raise ValueError("Password must contain at least one uppercase letter")
    if not _LOWERCASE_RE.search(value):
        raise ValueError("Password must contain at least one lowercase letter")
    if not _DIGIT_RE.search(value):
        raise ValueError("Password must contain at least one digit")
    if not _SYMBOL_RE.search(value):
        raise ValueError("Password must contain at least one symbol")
    return value

File: frontend/src/utils/passwordPolicy.ts
import { z } from 'zod'

// Mirrors backend/app/core/password_policy.py exactly: at least one
// uppercase letter, one lowercase letter, one digit, and one symbol, on
// top of the 8-72 char length bound. Kept in one place so Register, admin
// user create/edit, and password reset never drift from what the backend
// actually enforces — showing an inline error here that the server would
// reject anyway is worse than showing the real rule up front.
const UPPERCASE_RE = /[A-ZÁÉÍÓÚÑÜ]/
const LOWERCASE_RE = /[a-záéíóúñü]/
const DIGIT_RE = /\d/
// Anything that isn't a letter, digit, or whitespace — matches the
// backend's broad `[^\w\s]` symbol check rather than a fixed ASCII set.
const SYMBOL_RE = /[^\p{L}\p{N}\s]/u

export function passwordSchema(message: { tooShort: string; tooLong: string; weak: string }) {
  return z
    .string()
    .min(8, message.tooShort)
    .max(72, message.tooLong)
    .refine((v) => UPPERCASE_RE.test(v), message.weak)
    .refine((v) => LOWERCASE_RE.test(v), message.weak)
    .refine((v) => DIGIT_RE.test(v), message.weak)
    .refine((v) => SYMBOL_RE.test(v), message.weak)
}

export function isPasswordStrongEnough(value: string): boolean {
  return (
    value.length >= 8 &&
    value.length <= 72 &&
    UPPERCASE_RE.test(value) &&
    LOWERCASE_RE.test(value) &&
    DIGIT_RE.test(value) &&
    SYMBOL_RE.test(value)
  )
}
